경북대 FE 장세현 2단계 - 주문하기 페이지 구현하기

SeHIgh opened the pull request:
안녕하십니까 경북대학교 프론트엔드 과정 수강중인 장세현입니다.
2단계 과제 - 주문하기 페이지 구현하기 내용을 PR 요청 드립니다!

이번에는 진행 사항 정리 + 추가 구현 사항으로 정리 해보았습니다.

추가로 이번 PR 요청에는 작업 내용이 좀 많은 것 같아서 보다 더 제 의도를 잘 전달 드리기위해 질문 사항 및 구체적인 내용은 작성한 코드에 코멘트로 달아두었습니다!

---

### 💫 진행 사항 정리
- [x] (STEP1 피드백 반영) 유효성 검증 로직을 따로 분리하여 구현하기
- [x] (STEP1 피드백 반영) 유효성 검증 관련 정규식을 상수 파일로 따로 관리하기 -> 정규식 만이 아닌 유효성 검증을 위한 규칙 데이터 전체를 따로 통합 관리

- [x] 별도의 Form, Validation 라이브러리를 사용하지 않고 useState로만 구현하기

- 로그인 정보
  - [x] 로그인 정보를 어디서든 사용할 수 있게 하기
    - 필요한 로그인 정보 분석
      - 이름: 이메일 '@' 앞에 위치
      - 이메일 주소: ID 와 동일
  - [x] 새로고침을 하더라도 로그인 정보가 유지되도록 하기
  - [x] 로그인 한 유저는 로그인 버튼 클릭 시 마이페이지(/my)로 연결
  - [x] 마이페이지는 로그인 한 사람만 접근 가능. 로그인 하지 않은 사람이 접속하면 로그인 페이지로 연결
  - [x] 마이페이지에서 로그아웃 가능. 로그아웃 시 로그인(/login) 페이지로 연결

- 주문하기
  - [x] 상품 아이템을 클릭하면 주문하기 페이지로 연결
  - [x] 만약 로그인이 되어있지 않다면 로그인을 유도하고, 로그인 성공 시 주문하기 페이지로 접근 가능하게 구현
  - [x] 참고 URL을 살펴보고 주문하기에 필요한 기능을 구현
    - [x] 선물하기 카드 선택 및 문구 입력 (각 카드의 문구 기본값 + 수정 가능)
    - [x] 보내는 사람 정보 입력폼 (이름)
    - [x] 받는 사람 정보 입력 폼 (이름, 전화번호, 수량)
    - [x] 상품 정보
    - [x] 주문하기 버튼 (ex. 29000원 주문하기)
  - [x] 버튼 클릭 시 아래 유효성 검사 조건이 충족되지 않으면 각 폼 필드에 안내 메시지를 보여주기
    - 메시지는 반드시 입력 되어야 함
    - 보내는 사람 이름이 반드시 입력 되어야 함
    - 받는 사람 이름이 반드시 입력 되어야 함
    - 받는사람 전화번호가 반드시 입력되고 전화번호 규칙에 맞아야 함 (01012341234)
    - 수량은 1개 이상이어야 함

- [x] 본인만의 기준으로 일관된 코드 작성 하기
- [x] 기능 단위로 나누어 커밋 작성 하기
- [x] 재사용 가능성이 높은 컴포넌트, Hooks가 보이면 재사용성을 고려하여 만들기

---

### 🎯 추가 구현 사항 및 개선점
- 에러 메시지 표시 타이밍을 최적화하였습니다.
  - 로그인 폼과 달리 blur 시가 아닌 주문하기 제출 시에만 표시됩니다.
- 각 섹션별로 독립적인 컴포넌트 설계하여 유지보수성을 향상 시켰습니다.
- 시맨틱 HTML 태그를 활용하였습니다. (section, label 등)

SeHIgh committed:
docs(README): 2단계 과제 요구사항 정리 및 1단계 과제 피드백 반영 추가 기능 정리

SeHIgh committed:
feat(validate): 유효성 검증 로직 분리 및 구현 & 규칙 데이터 통합 관리 파일 생성

- 기존 정규식 만 분리하여 관리하기로 했던 목표 -> 정규식 만이 아닌 유효성 검증을 위한 규칙 데이터 전체를 따로 통합 관리 하도록 개선
- 관련 목표 내용 README 파일 수정

- 유효성 검사를 위한 규칙 Rule 타입 생성
- Rule 타입은 4가지를 적을 수 있고, 각 필드 별 value 와 errorMsg(에러메시지)를 따로 넘겨주도록 설정
- 에러 메시지는 없을 수 있다 가정하에 기본 반환 값 설정

- 유효성 검사 유틸함수에 맞게 기존 useInput 커스텀 훅 및 Login 페이지 수정

SeHIgh committed:
feat(auth): Context API 기반 로그인 정보 전역 상태 관리 구현

- 사용자의 로그인 정보를 전역적으로 관리하는 AuthContext.tsx 컨텍스트 구현 (Context API 기반)
- 새로고침 시에도 로그인 정보가 유지되도록 쿠키를 이용하여 저장
- 로그인 정보 저장 및 사용 시 보안을 위해 암호화/복호화 유틸 함수를 구현 및 사용 하여 쿠키에 저장

SeHIgh committed:
feat(auth): 구현한 인증 정보 컨텍스트 적용 및 테스트

- App.tsx 에 AuthProvider 적용
- Login.tsx 에 이전에 구현된 로그인 로직에 인증 정보 저장하도록 적용
- Friends.tsx 선물할 친구 선택 섹션에 로그인 시 사용자 이름을 추가 렌더링하도록 개선
- getCookie에서decodeURIComponent로 디코딩하지 않고 바로 반환했던 문제 수정

SeHIgh committed:
feat(auth): 마이페이지 구현, 라우팅 및 로그인 상태에 따른 네비게이션 개선

- Router.tsx에 마이페이지 라우트 추가 및 리다이렉트 구현
- Button.tsx에서 기본 버튼 변형 추가 및 스타일 개선
- Nav.tsx에서 로그인 상태에 따라 네비게이션 버튼 동작 변경
- Login.tsx에서 로그인 후 이전 페이지로 리다이렉트 기능 추가
- README.md에 로그인 정보 사용 및 유지 관련 항목 체크 완료 표시

SeHIgh committed:
refactor(validate): 유효성 검사 로직 단순화 및 UX 개선

- useInput 훅의 복잡한 상태 관리 로직 단순화
- 초기 상태에서 input border 빨간색 표시 문제 해결
- Login 페이지 버튼 활성화 조건 최적화

SeHIgh committed:
feat(hook): reset 함수 추가로 입력 상태 초기화 기능 구현

SeHIgh committed:
docs(order-mock): 주문하기 카드 템플릿 목데이터 추가

SeHIgh committed:
feat(order): 주문하기 페이지 및 라우팅 추가, 상품 클릭 시 로그인 상태에 따른 페이지 이동 구현

- ROUTH_PATH 에 주문하기 페이지 경로 추가
- 각 상품에 맞는 동적 페이지를 위해 동적 라우트 정의
- 주문하기 페이지 구현
- 상품 아이템 컴포넌트에서 상품 클릭시 핸들러 작성
- 추후 id + 1 을 지속적으로 쓰기보다 MOCK 데이터의 id를 index + 1 로 설정하여 불필요한 코드 개선

SeHIgh committed:
docs(README, order): 주문하기 페이지의 체크리스트 업데이트 및 구현 할 세부기능 정리

SeHIgh committed:
feat(order-card): 주문 카드 섹션 구현 및 주문 페이지 업데이트

- 주문 페이지에서 카드 선택 및 메시지 입력을 처리하는 컴포넌트 구현
- 주문 페이지를 업데이트하여 OrderCardSection을 포함하고, 사용하지 않는 Typography 컴포넌트를 제거
- placeholder 색상 설정을 추가
- 입력 필드를 위한 타이포그래피 스타일을 생성
- 카드 리스트, 썸네일, 메시지 영역을 포함한 OrderCardSection의 스타일 컴포넌트 제작

SeHIgh committed:
feat(order): 주문하기 페이지 구현

- 주문 폼 컴포넌트 구현 (카드, 보내는/받는 사람 정보)
- 유효성 검증 시스템 적용 (제출 시에만 에러 표시)
- 상품 정보 및 할인률 표시 로직 구현
- Typography/Button 컴포넌트로 일관된 UI 적용

SeHIgh commented on src/components/home/Trend/ProductItem.tsx:
```diff
@@ -2,12 +2,30 @@ import { theme } from '@/styles/theme'
+      navigate(`${ROUTH_PATH.ORDER}/${product.id}`)
+    } else {
+      // * 로그인 (무) -> 로그인 페이지 (로그인 이후 from을 남겨 )
+      navigate(ROUTH_PATH.LOGIN, { state: { from: `${ROUTH_PATH.ORDER}/${product.id}` } })
```
현재는 상품을 눌러 주문하기 페이지로 이동 시 로그인이 안되어있다면, 로그인화면으로 이동 하고 이때, from state를 넘겨주어 로그인 이후 이동해야할 경로를 지정해주는 형태로 설정하였는데, 이런 방식이 적절한지 궁금합니다!

terajh replied:
당연히 정답은 없기 때문에, 충분한 논리와 기준을 바탕으로 작성해주셨다면 적절한 구현이라고 생각합니다! 🙌 다만 현재 코드 기준으로는 url 입력을 통해 주문하기 페이지에 직접 접근했을 때 로그인 페이지로 이동시켜주지 않을 것 같아요. 이를 참고해 고려해볼 만한 방식이 있어 공유드려볼게요 :)

현재 로그인 유무에 따라 접근 가능한지 판단하는 로직은 주문하기 페이지뿐 아니라 다른 여러 페이지에서도 동일하게 필요해질 수 있는 부분인데요. ProductItem 에서 클릭이벤트에만 로그인 여부 확인 로직을 넣어줄 경우, 위에서 적은 것처럼 페이지 url 로 직접 접근시에는 뚫릴 수 있어보여요.

이럴 경우, 해당 로직을 HOC(High Order Component) 형태로 분리해두면 여러 페이지에서 재사용하기 좋고, 인증 로직도 한 곳에서 관리할 수 있게 책임을 부여할 수 있어 비교적 안전하게 사용할 수 있을 것 같네요. 아래는 간단한 예시인데요. 수정하실때 참고차 남겨볼게요ㅎㅎ

```ts
// hoc/withAuth.tsx
export function withAuth(ChildrenComponent) {
  return function AuthProtectComponent(props) {
    const { user } = useAuth();
    const location = useLocation();

    if (!user) {
      return <Navigate to="/login" state={{ from: location.pathname }} replace />;
    }

    return <ChildrenComponent {...props} />;
  };
}

// pages/Order
import { withAuth } from '@/hoc/withAuth';
function Order() {
  return ....
}

export default withAuth(Order);

```

SeHIgh replied:
헉, url로 접속한다는 부분을 생각하지 못했었습니다.
멘토님 께서 제안해주신 HOC 패턴을 공부하고, 구현하여 코드 리뷰에 반영 하였습니다!
꼼꼼하게 리뷰해주셔서 감사합니다!
구현한 HOC을 이용하여 이후에 구현 할 새로운 페이지들에 인증 로직을 편하게 적용할 수 있을 것 같습니다.

SeHIgh commented on src/components/home/Trend/Trend.tsx:
```diff
@@ -101,8 +101,8 @@ export const Trend = () => {
       <ProductContainer>
-        {displayProducts.map((product, index) => (
-          <ProductItem key={product.id} product={product} index={index} />
+        {displayProducts.map((product) => (
```
상품 id 자체가 Mock data에 각 상품의 고유 값으로 순서 +1 로 값이 들어가도록 하여 '순위 == 랭크' 가 되도록 바꾸었기 때문에 굳이 index를 넘길 필요가 없다 판단하여 넘겨주는 값에서 빼게되었습니다.

근데 코멘트를 달면서 다시 보니 현재 상황에서는 문제가 되지않지만, 
일반적으로는 상품 고유 id는 보통 순위가 아니기도 하고, 특정 기준에 따라 필터링을 통해서 순위가 달라질 수 있기 때문에 이는 옳지 않은 코드 같다 판단이 됩니다.

일정상 수정이 불가능하여 다음 리뷰에 수정하도록 하겠습니다...!

SeHIgh commented on src/components/order/OrderCardSection.tsx:
```diff
@@ -0,0 +1,212 @@
+}
+
+// * 섹션 컨테이너 : section 시맨틱 태그
+const SectionContainer = styled.section`
```
이전에 SEO를 위해 시멘틱 태그를 사용하는게 좋다고 배웠어서 이번 섹션에 섹션 시멘틱 태그 적용하여 보았습니다!
(이후에 이전에 작성한 다른 섹션 컴포넌트들도 시멘틱 태그로 전체 바꿔줄 예정입니다!)

SeHIgh commented on src/contexts/AuthContext.tsx:
```diff
@@ -0,0 +1,71 @@
+    setUser(userInfo)
+    setIsLogin(true)
+    // ! 쿠키에 하루 동안 저장
+    setCookie('userInfo', encodeUserInfo(userInfo), 1)
```
처음에는 세션스토리지를 사용할까 생각도 하였지만, 보안을 위하여 실제 로그인 형태에서 사용되는 쿠키를 이용하여 로그인 정보를 저장하였습니다.

terajh replied:
좋은 수정이라고 생각합니다! 👍👍 실제 업무로 코드 작성시에는 개발간의 컴을 통해 로그인 정보를 저장하는 쿠키값 저장 기한을 정해두고 사용하곤하는데요. 참고차 코멘트 남깁니다ㅎㅎ

SeHIgh commented on src/contexts/AuthContext.tsx:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  // * 로그아웃 함수
+  const logout = () => {
```
로그 아웃 시에 안전하게 로그인 정보 상태를 초기화하고, 쿠키도 삭제하도록 하였습니다!

SeHIgh commented on src/data/validateRules.ts:
```diff
@@ -0,0 +1,66 @@
+// ! 필드명을 고정(한정)하고 싶지 않아 string 타입으로 두었는데,
+// ! 이 경우 단순히 Record<> 로 두면 잘못된 필드명 사용 시 걸러지지 않는 문제 발생
+// ! 따라서, as const satisfies Record<> 구문을 이용하여 타입 유도
+export const VALIDATE_RULES = {
+  email: {
+    required: { value: true, errorMsg: 'ID를 입력해주세요.' },
+    regex: {
+      value: /^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$/,
+      errorMsg: 'ID는 이메일 형식으로 입력해주세요.',
+    },
+  },
+  password: {
+    required: { value: true, errorMsg: 'PW를 입력해주세요.' },
+    minLength: { value: 8, errorMsg: 'PW는 최소 8글자 이상이어야 합니다.' },
+  },
+  // * 주문하기 관련 유효성 검증 규칙
+  message: {
+    required: { value: true, errorMsg: '메시지를 입력해주세요.' },
+  },
+  name: {
+    required: { value: true, errorMsg: '이름을 입력해주세요.' },
+  },
+  phone: {
+    required: { value: true, errorMsg: '전화번호를 입력해주세요.' },
+    regex: {
+      value: /^010\d{8}$/,
+      errorMsg: '올바른 전화번호 형식이 아닙니다.',
+    },
+  },
+  quantity: {
+    required: { value: true, errorMsg: '수량을 입력해주세요.' },
+    custom: {
+      value: (value: string) => {
+        const num = Number(value)
+        return !isNaN(num) && num >= 1
+      },
+      errorMsg: '구매 수량은 1개 이상이어야 합니다.',
+    },
+  },
+} as const satisfies Record<string, Rule>
```
주석에도 적어두었지만, 유효성 검증에 대한 규칙 관리를 더 편하게 할 수 있도록 별도 데이터를 관리하도록 구현하였습니다.

여기서 추가로 타입 안전성을 위해 객체 타입 지정을 위해 Record를 사용하였습니다.
그런데, 필드명을 고정(제한)하고 싶지 않아서 string으로 두고 단순히 Record<> 형태로 두었더니 관리되지 않는 잘못된 필드명 사용 시에도 걸러지지 않는 문제가 발생하였습니다.
그래서 이를 방지하기 위해 as const satisfies 구문을 이용하여 현재 관리 되는 필드명으로 타입이 유도되도록 하였습니다.

terajh replied:
[질문] 매우 좋은 방법이네요! 한가지 질문사항이 있는데요. 개인적인 제 생각에는 validate 대상들을 아래처럼 제한하게 되었을 때 VALIDATE_RULES 상수를 가져다 사용하는 측에서 안전하게 타입 추론이 되어 선언해두지 않은 key 를 사용하는 사이드 이펙을 막을 수 있지 않을까? 생각이 되어서요. 혹시 필드명을 고정하고 싶지 않다고 하셨는데, 혹시 관련해서 그렇게 생각하신 이유가 있으신걸까요??


```ts
type ValidateKey = 'email' | 'password' | message' | 'name' | 'phone' | 'quantity';
export const VALIDATE_RULES: Record<ValidateKey, Rule> = {
```

SeHIgh replied:
사실 저도 코드 작성을 위해 여러 자료를 찾아보니 멘토님께서 알려주신 방법 형태로 작성을 하는게 정석적인 방법인 것 같아 처음에는 그렇게 작성을 했었습니다.
그런데 유효성 검사를 따로 분리하고 규칙 데이터를 따로 둔 것 자체가 여러 부분에서 사용될 것이라는 생각으로 만들게 된 것인데,
해당 부분에 필드명을 타입 지정을 통해 제한해버리면, 새로운 필드명 추가시에 타입을 추가하고 규칙 데이터에도 추가하는 형태의 이중 작업이 요구 돼버리므로 사소하지만, 해당 함수를 개발자 입장에서 좀 더 편하게 사용 할 수 있는 방법을 적용하고자 이런식으로 구현하게 되었습니다!

SeHIgh commented on src/hooks/useInput.ts:
```diff
@@ -1,33 +1,41 @@
+    setValue(newValue)
 
-    // * 유효성 검사 함수가 있으면 실행
-    if (isBlurred && validate) {
-      const errorMessage = validate(e.target.value)
-      setError(errorMessage)
+    // * blur된 상태에서는 실시간으로 에러 업데이트
+    if (isBlurred) {
+      const newError = validateValue(newValue, rule)
+      setError(newError)
     }
   }
 
   // * Blur 시 유효성 검사 실행
   const handleBlur = () => {
     setIsBlurred(true)
-    if (validate) {
-      const errorMessage = validate(value)
-      setError(errorMessage)
-    }
+    const errorMessage = validateValue(value, rule)
+    setError(errorMessage)
+  }
+
+  // * 초기화
+  // ? 추후 여러 폼 작성 시에 필요할 것 같아 추가
+  const reset = () => {
+    setValue(initialValue)
+    setError(null)
+    setIsBlurred(false)
   }
 
-  // * 유효성 검사 결과
-  const isValid = !error
+  // * 유효성 여부 (버튼 활성화용)
+  const isValid = validateValue(value, rule) === null
 
-  return { value, error, handleChange, handleBlur, isValid }
+  return { value, error, handleChange, handleBlur, isValid, reset }
 }
```
따로 분리된 유효성 검사 유틸 함수를 전체적으로 적용하였습니다.

SeHIgh commented on src/pages/Login.tsx:
```diff
@@ -37,7 +53,7 @@ export const Login = () => {
           value={email.value}
           onChange={email.handleChange}
-          hasError={!email.isValid}
+          hasError={!!email.error}
```
보다 정확한 에러상태를 전달하기 위해 유효성 여부가 아닌 에러 상태의 유무를 이용하여 판단하도록 개선하였습니다.

SeHIgh commented on src/styles/typography.ts:
```diff
@@ -0,0 +1,14 @@
+import { css } from '@emotion/react'
+import { theme } from './theme'
+
+export const typographyInput = css`
+  font-size: ${theme.typography.body.body1Regular.fontSize};
+  font-weight: ${theme.typography.body.body1Regular.fontWeight};
+  line-height: ${theme.typography.body.body1Regular.lineHeight};
+`
+
+export const typographyLabel = css`
+  font-size: ${theme.typography.body.body1Regular.fontSize};
+  font-weight: ${theme.typography.body.body1Regular.fontWeight};
+  line-height: ${theme.typography.body.body1Regular.lineHeight};
+`
```
반복되어 사용되는 typography를 따로 구성하여 여러 곳에서 쉽게 사용할 수 있도록 개선하였습니다.

terajh replied:
크.. 재사용 여부에 따른 격리는 언제나 옳습니다 👍

SeHIgh commented on src/utils/encodingUtils.ts:
```diff
@@ -0,0 +1,16 @@
+import type { UserInfo } from '@/contexts/AuthContext'
+
+// * 유저 정보 암호화
+// ? 조금이라도 보안성을 주기 위해 간단한 Base64 인코딩 적용
+export const encodeUserInfo = (userInfo: UserInfo) => {
+  return btoa(JSON.stringify(userInfo))
+}
+
+// * 유저 정보 복호화
+export const decodeUserInfo = (encodedInfo: string): UserInfo | null => {
+  try {
+    return JSON.parse(atob(encodedInfo))
+  } catch {
+    return null
+  }
+}
```
조금이라도 보안성을 위해 유저 정보를 암호화/복호화 하여 저장 할 수 있도록 유틸함수를 작성하였습니다.

terajh replied:
보안까지 생각해주시다니 좋네요 👍

SeHIgh commented on src/utils/encodingUtils.ts:
```diff
@@ -0,0 +1,16 @@
+import type { UserInfo } from '@/contexts/AuthContext'
+
+// * 유저 정보 암호화
+// ? 조금이라도 보안성을 주기 위해 간단한 Base64 인코딩 적용
+export const encodeUserInfo = (userInfo: UserInfo) => {
+  return btoa(JSON.stringify(userInfo))
+}
+
+// * 유저 정보 복호화
+export const decodeUserInfo = (encodedInfo: string): UserInfo | null => {
+  try {
+    return JSON.parse(atob(encodedInfo))
+  } catch {
+    return null
+  }
+}
```
조금이라도 보안성을 위해 유저 정보를 암호화/복호화 하여 저장 할 수 있도록 유틸함수를 작성하였습니다.

SeHIgh commented on src/utils/validateValue.ts:
```diff
@@ -0,0 +1,44 @@
+/**
+ * * value에 대해 여러 유효성 검사를 수행하는 유틸 함수
+ */
+export const validateValue = (value: string, rules: Rule): string | null => {
+  const trimmedValue = value.trim()
+
+  if (rules.required && !trimmedValue) {
+    return rules.required.errorMsg || '필수 입력 항목입니다.'
+  }
+
+  if (rules.regex && !rules.regex.value.test(trimmedValue)) {
+    return rules.regex.errorMsg || '형식이 올바르지 않습니다.'
+  }
+
+  if (rules.minLength && trimmedValue.length < rules.minLength.value) {
+    return rules.minLength.errorMsg || `${rules.minLength}자 이상 입력해주세요.`
+  }
+
+  if (rules.custom && !rules.custom.value(trimmedValue)) {
+    return rules.custom.errorMsg || '유효하지 않은 입력입니다.'
+  }
+
+  return null
+}
```
이전 유효성 검사를 따로 분리 하면 좋겠다는 피드백을 반영하여 해당 유효성 검사 유틸함수를 따로 만들어 보았습니다.

해당 유틸함수는 따로 관리되는 유효성 검증 규칙 데이터에 따라 많이 사용되는 형태의 유효성 검증 3가지 + 사용자 정의 검사, 총 4가지의 유효성 검증을 진행 할 수 있도록 구성하였습니다.

terajh replied:
오 매우 깔끔하게 분리해주셨네요. 여러 곳에서 유용하게 사용할만한 유틸이라고 생각합니다. 👍

SeHIgh commented on src/utils/validateValue.ts:
```diff
@@ -0,0 +1,44 @@
+// * 유효성 검증을 위한 규칙 타입
+export type Rule = {
+  required?: {
+    value: boolean
+    errorMsg?: string
+  }
+  regex?: {
+    value: RegExp
+    errorMsg?: string
+  }
+  minLength?: {
+    value: number
+    errorMsg?: string
+  }
+  custom?: {
+    value: (value: string) => boolean
+    errorMsg?: string
+  }
+}
+
```
유효성 검사에는 4가지 타입의 규칙을 전달 받을 수 있도록 구성하였고, 유지보수성을 위해 각 규칙에 따라 에러메시지도 같이 관리 할 수 있도록 하였습니다.

terajh commented on src/pages/Order.tsx:
```diff
@@ -0,0 +1,320 @@
+  // * 폼 유효성 검사
+  const validateForm = () => {
+    const errors = {
+      cardMessage: validateValue(cardMessage, VALIDATE_RULES.message),
```
유효성 로직을 격리시켜주어서 동일한 로직이 필요한 곳에서 깔끔하게 코드 작성을 할 수 있게 되었네요 👍👍

terajh reviewed (changes requested):
안녕하세요 세현님! 2단계 미션 고생 많으셨습니다.

전반적으로 2단계 미션의 요구사항들을 잘 구현해주신것 같아요. 커밋단위도 진짜 깔끔하게 잘 나눠주셨구요. 컴포넌트 뿐 아니라 데이터상수 분리, 재사용성 및 보안을 고려한 유틸도 잘 구현해주셔서 너무 깜짝놀랐네요ㅎㅎ 고생 많으셨습니다. 👏🏻👍🏻

추가로, 주문하기 페이지에 직접 진입할 경우 로그인 페이지로 리다이렉트되지 않는 이슈에 대해 의견을 남겨두었습니다. 관련해 참고하실 만한 방법들을 코멘트로 함께 정리해두었으니, 확인하시고 한 번쯤 고민해보시면 좋을 것 같아요 🙃 해당 부분이 현재 기능 구현에 중요한 포인트라 판단되어 request changes로 남겨드렸습니다.

과제 진행 중 고민했던 부분이나 궁금한 점이 있으시면, 언제든 PR description이나 DM으로 편하게 남겨주세요! 감사합니다! 🙏🏻

SeHIgh committed:
fix(trend): 상품 아이템에 순위 표시 로직 변경 id -> index

SeHIgh committed:
fix(route): ROUTE_PATH 철자 수정 (ROUTH->ROUTE) 및 전체 반영

SeHIgh committed:
feat(auth): 로그인이 요구되는 페이지를 위한 HOC 추가 및 라우터 업데이트

- 인증 처리 로직을 위한 HOC 패턴 구현
- 라우터에서 별도로 인증 HOC을 적용할 페이지들 통합 관리
- MyPage와 Order 페이지에 HOC 패턴을 고려하여 인증 처리 로직 제거
- ProductItem 컴포넌트에서 로그인 상태에 따른 페이지 이동 로직 제거
- MyPage에서 로그인 상태 체크 로직 제거 및 리다이렉트 처리 개선

SeHIgh committed:
refactor(SEO): 검색엔진 최적화를 위한 시맨틱 태그 적용
- main, section, article

SeHIgh commented on src/Router.tsx:
`ROUTH` 철자가 잘못되어있는 것을 발견하여 `ROUTE` 로 철자를 수정하여 다른 코드들에도 전체 반영하였습니다.
또한, 새롭게 구현된 `WithAuth()` HOC 을 Router페이지에서 인증이 필요한 페이지들을 별도 객체로 관리하도록 적용하였습니다.

SeHIgh commented on src/components/home/Trend/ProductItem.tsx:
```diff
@@ -31,8 +42,8 @@ export const ProductItem = ({ product, index }: { product: Product; index: numbe
-// * 실시간 급상승 상품 아이템 컨테이너
-const ProductItemContainer = styled.div`
+// * 실시간 급상승 상품 아이템 컨테이너 (article 시맨틱 태그 사용)
+const ProductItemContainer = styled.article`
```
반복적으로 사용되는 상품 아이템은 `article` 시멘틱 태그를 적용해보았습니다.

SeHIgh commented on src/hoc/withAuth.tsx:
멘토님 께서 제안해주신 HOC 패턴을 공부하고, 코드를 참고하여 타입을 추가하여 구현해보았습니다.
그런데, `ChildrenComponent` 를 반환하는 부분에 어떻게 해도 타입 오류가 지속적으로 발생하여 `as any` 를 사용하고 해당 오류를 무시하도록 문구를 추가하였습니다.

SeHIgh replied:
지금 구현된 코드와 같이 any를 쓰는 것은 지양해야 한다 알고 있어서
혹시 해당 코드를 오류없이 개선할 수 있는 방법이 있는지 궁금합니다.

terajh replied:
any는 가능한 피하는 것이 좋다는 건 다들 잘 아실 거예요. 다만, 이번 props의 경우 살펴보니 React HOC 패턴에서는 타입 시스템의 한계로 인해 props 타입을 완전히 명시적으로 추론하기 어려운 상황이 있는 것 같네요.

런타임에서 문제가 없다는 게 확실하고, 타입 시스템으로 이를 깔끔하게 표현하기 어려운 경우라면, 현재처럼 주석 등을 통해 "의도적으로 any를 사용했음"을 명시해주는 방식도 괜찮다고 생각합니다. 다만, 모든 상황에서 주석을 통해 any를 사용하는 것 보단 가능한 경우엔 unknown이나 제네릭 등을 활용해 타입 안정성을 조금이라도 확보하는 방법도 괜찮겠네요! 고생하셨습니다ㅎㅎ

SeHIgh reviewed (commented):
멘토님 께서 주셨던 피드백 및 계획했던 수정 내용을 반영하였습니다!

### 💫 피드백 반영 수정
- 상품 아이템에 순위 표시 로직을 다시 이전 형태로 고쳤습니다.
  - `id -> index`
- 철자가 잘못 된 부분을 발견하여 수정 및 전체 반영 하였습니다.
  - `ROUTH_PATH` -> `ROUTE_PATH`
- 멘토님의 피드백을 반영하여 인증 HOC 구현 및 해당 HOC 을 `Router.tsx`에서 인증(로그인)이 요구되는 페이지들을 통합 관리하도록 개선 하였습니다.
  - 추가로 인증 HOC 구현으로 인해 Order 및 Mypage에서 중복으로 사용되는 리다이렉트 코드를 제거하였습니다.
- 이전에 적용한 시멘틱 태그를 다른 부분들에도 전체적으로 리팩토링 하였습니다.
  - `section` 태그 뿐만아니라 `main`, `article` 도 적용해보았습니다.

terajh reviewed (approved):
코멘트 관련해 수정후 PR 요청 주신걸로 확인되어 승인 후 머지처리하도록 하겠습니다. 고생하셨습니다~!
